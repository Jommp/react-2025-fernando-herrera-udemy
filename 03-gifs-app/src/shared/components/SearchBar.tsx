import { useEffect, useState, type KeyboardEvent } from 'react';

interface Props {
  placeholder: string;
  button?: string;
  onSearch: (query: string) => void;
};

export const SearchBar = ({ placeholder, button = 'Buscar', onSearch }: Props) => {
  const [query, setQuery] = useState('');

  useEffect(() => {
    const timeOutId = setTimeout(() => {
      handleSearch();
    }, 900);
  
    return () => {
      clearInterval(timeOutId);
    }
  }, [query, onSearch]);
  
  const handleSearch = () => {
    onSearch(query);
  };

  const handleKeyDown = (key: KeyboardEvent<HTMLInputElement>) => {
    if (key.key !== 'Enter') return;

    handleSearch();
  };

  return (
    <section className='search-container'>
      <input
        type='text'
        placeholder={ placeholder }
        onChange={(ev) => setQuery(ev.target.value)}
        onKeyDown={handleKeyDown}
      />

      <button onClick={handleSearch}>
        { button }
      </button>
    </section>
  );
};
