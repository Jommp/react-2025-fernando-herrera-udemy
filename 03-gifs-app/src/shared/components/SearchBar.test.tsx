import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';

import { SearchBar } from './SearchBar';

describe('SearchBar', () => {
  test('should render SearchBar correctly', () => {
    const { container } = render(<SearchBar placeholder={'Buscar'} onSearch={() => {}}/>);

    expect(container).toMatchSnapshot();

    expect(screen.getByRole('textbox')).toBeDefined();
    expect(screen.getByRole('button')).toBeDefined();
  });

  test('should call onSearch with the correct value after 900ms', async () => {
    const onSearch = vi.fn();
    render(<SearchBar placeholder={'Buscar'} onSearch={onSearch}/>);

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'test' } });

    await waitFor(() => {
      expect(onSearch).toHaveBeenCalled();
      expect(onSearch).toHaveBeenCalledWith('test');
    });
  });

  test('should call only once with the last value', async () => {
    const onSearch = vi.fn();
    render(<SearchBar placeholder={'Buscar'} onSearch={onSearch}/>);

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 't' } });
    fireEvent.change(input, { target: { value: 'te' } });
    fireEvent.change(input, { target: { value: 'tes' } });
    fireEvent.change(input, { target: { value: 'test' } });

    await waitFor(() => {
      expect(onSearch).toHaveBeenCalledTimes(1);
      expect(onSearch).toHaveBeenCalledWith('test');
    });
  });

  test('should call onSearch when button is clicked with the input value', () => {
    const onSearch = vi.fn();
    render(<SearchBar placeholder={'Buscar'} onSearch={onSearch}/>);

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'test' } });

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith('test');
  });

  test('should the input has the correct placeholder value', () => {
    const placeholder = 'Buscar Gif';

    render(<SearchBar placeholder={placeholder} onSearch={() => {}}/>);

    expect(screen.getByPlaceholderText(placeholder)).toBeDefined();
  });
});
