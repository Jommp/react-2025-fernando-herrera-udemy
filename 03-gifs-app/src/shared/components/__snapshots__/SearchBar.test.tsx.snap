// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > should render SearchBar correctly 1`] = `
<div>
  <section
    class="search-container"
  >
    <input
      placeholder="Buscar"
      type="text"
    />
    <button>
      Buscar
    </button>
  </section>
</div>
`;
